pkg/puller/downloader.go

```diff
@@ -0,0 +1,173 @@
+package puller
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+var (
+	d *Downloader
+)
+
+type Downloader struct {
+	NumRetries int
+	ModelDir   string
+}
+
+type Protocol string
+
+const (
+	// Supported Protocols
+	S3    Protocol = "s3://"
+	GS    Protocol = "gs://"
+	PVC   Protocol = "pvc://"
+	File  Protocol = "file://"
+	HTTPS Protocol = "https://"
+)
+
+// TODO: Add more supported protocols
+var SupportedProtocols = []Protocol{S3}
+
+func init() {
+	d = NewDownloader()
+}
+
+func NewDownloader() *Downloader {
+	d := new(Downloader)
+	return d
+}
+
+func SetEnvs(numRetries int, modelDir string) {
+	d.NumRetries = numRetries
+	d.ModelDir = modelDir
+}
+
+func DownloadModel(event EventWrapper) error {
+	modelSpec := event.ModelSpec
+	if modelSpec != nil {
+		modelUri := modelSpec.StorageURI
+		hashString := hash(modelUri)
+		log.Println("Processing:", modelUri, "=", hashString)
+		successFile := filepath.Join(d.ModelDir, fmt.Sprintf("SUCCESS.%d", hashString))
+		ok := fileExists(successFile)
+		if !ok {
+			downloadErr := download(modelUri)
+			if downloadErr != nil {
+				return fmt.Errorf("download error: %v", downloadErr)
+			} else {
+				file, createErr := os.Create(successFile)
+				if createErr != nil {
+					return fmt.Errorf("create file error: %v", createErr)
+				}
+				defer file.Close()
+			}
+		} else {
+			log.Println("Model", modelSpec.StorageURI, "exists already")
+		}
+	}
+	return nil
+}
+
+func download(storageUri string) error {
+	log.Println("Downloading: ", storageUri)
+	prefix, err := validateStorageURI(storageUri)
+	if err != nil {
+		return err
+	}
+	switch prefix {
+	case S3:
+		s3Uri := strings.TrimPrefix(storageUri, string(S3))
+		path := strings.SplitN(s3Uri, "/", 2)
+		bucket := path[0]
+		item := path[1]
+		log.Println("do an s3 request on", bucket, "for key", item)
+		fileName := filepath.Join(d.ModelDir, item)
+		ok := fileExists(fileName)
+		if ok {
+			// File got corrupted or is mid-download :(
+			// TODO: Figure out if we can maybe continue?
+			log.Println("Deleting", fileName)
+			err := os.Remove(fileName)
+			if err != nil {
+				return fmt.Errorf("file is unable to be deleted: %v", err)
+			}
+		}
+		file, err := create(fileName)
+		if err != nil {
+			return fmt.Errorf("file is already created: %v", err)
+		}
+		defer file.Close()
+
+		// TODO: Should the S3 client be shared?
+		sess, err := session.NewSession()
```

Comment:
According to https://docs.aws.amazon.com/sdk-for-go/api/aws/session/ each session refers to one set of access keys. If one InferenceServce's models all come from the same s3 tenancy we should be able to share it.

Reply:
Yeah, I am just unsure if it can handle concurrency. i.e. multiple go-routines calling the same session. For now, I will just leave it as is, if that is okay.

Reply:
For the record ```
Sessions are safe to use concurrently as long as the session isn't being modified. The SDK doesn't modify the session once the session is created. Creating service clients concurrently from a shared session is safe.
```